### DISCUSSION TITLE
RFC-2: sbt 2.0 RFC process

### DISCUSSION THREAD

**Discussion by eed3si9n** [sbt/sbt#7188] [Ideas]
https://eed3si9n.com/sbt-2.0-rfc-process

**Comment by ckipp01** [comment]
I wanted to just throw this in here as an adjacent thing to RFC. It's something we've adopted in some of our #work projects that I've come to really appreciate, especially when coming to a project after decisions have been made. I wanted to mention it during the summit sessions, but we ran out of time. In the article you mention:

> “a lightweight process based on RFC (request for comments) docs that can serve as a design document.” In general, it would be a good idea to capture motivations and design intent of major changes.

RFC's work great for feedback and discussion around upcoming changes and decisions, but don't always do the best job succinctly summarizing the decision. I often feel this way about SIP, RFCs, Contributor discussions, etc that we literally can have hundreds of comments that someone may need to comb through to get the full picture of why a decision is made. One way around this is to ensure that _when_ the RFC/SIP/whatever has finished that an [Architectural Decision Record](https://adr.github.io/) or ADR is created summarizing why the decision was made. In the projects I've used them in before we literally keep these in a `doc/architecture/decisions/` dir. This may be overkill and it _does_ add more maintenance burden, but I've personally found that the pros of having these types of decisions documented and summarized really help newcomers quickly understand architectural  decisions and why they were made without having to wade through huge threads of conversations. So long-run the additional burden seems to be worth it.

So interpret this as an additional _nice to have_ as we consider an RFC process. Taking the results of these RFCs and summarizing them somehow in a shared place would be great.

Thanks for all the work you're doing making sure 2.0 is a success!

**Reply by eed3si9n** [reply]
Yea, unlike Mozilla folks who maintain Rust, maintaining sbt isn't my job, so I don't think we should add too much work. Otherwise it would discourage me from writing quick memos if I need to follow up with summary of discussion doc etc.

**Reply by eed3si9n** [reply]
Maybe a lighter-weight version is that we should add some "outcome" section at the end of the doc when we accept or reject an RFC?

**Reply by ckipp01** [reply]
> Maybe a lighter-weight version is that we should add some "outcome" section at the end of the doc when we accept or reject an RFC?

Yea that would totally be fine.

**Comment by dragos** [comment]
@eed3si9n thank you for putting out the first RFC! I totally agree with the cited goals. 

On the specifics, I am used to Google Docs, which have the nice feature that discussion threads are linked to specific sections and can be independently "Resolved". This also makes it easier to "catch up" when arriving late to a discussion, so I'd suggest it as a default option.

I wonder if we can do away with "status". Since there is no Sbt committee, there is nobody with the authority to "approve". If the author of an RFC issues a PR, the PR status would tell us if it's implemented or if there are serious concerns.

**Reply by eed3si9n** [reply]
Re: status, even though for now I'd be proposing and approving them mostly on my own, it's important for any RFC to have some sort of a finite-state-machine. Otherwise, it's just memo + endless debate. For example, RFC-2 can be modified during `Review` period, then `Accepted` (end-of-review), or if deemed not a good idea, be be rejected or put on hold, and go straight to `Closed`.

**Reply by eed3si9n** [reply]
Re: Google Docs, I generally _love_ Google docs myself for internal memos, but for public RFCs there are some issues:
- It's sort of distracting to read the original doc when there are too many comments. This is more so true for public RFC where anyone can comment on anything with wide spectrum of quality/opinions.
- On the discussion side, the commenting is too narrow to have any meaningful discussion. See RFC-3 discussion (https://github.com/sbt/sbt/discussions/7189) for example where people are writing paragraphs of comments and equally lengthy (and interesting) replies. For internal stuff, we'd open up Edit rights maybe, but for now discussion is working out?

**Reply by SethTisue** [reply]
I tend to agree that Google Docs is excellent for collaboratively assembling and improving a piece of writing, but really not so good for debating and discussing issues.

**Comment by eed3si9n** [comment]
Hopefully we got all the inputs? I am accepting the RFC.

**Comment by ScalaWilliam** [comment]
Is there any chance to place more focus on fixing existing bugs and removing warts, in preference to embarking on redesign?

https://github.com/sbt/sbt/issues/6101#issuecomment-1491095989
https://github.com/sbt/sbt/issues?q=is%3Aissue+is%3Aopen+label%3ABug

The user experience is still not smooth in many areas, the documentation is also lacking.

I am also looking at this page, https://eed3si9n.com/sbt-2.0-rfc-process - the only place this page links to is to this GitHub Issue, and the only place this GitHub Issue links is that page. 😕  - not very sure how to proceed

---

When we look at the docs, it says that the sbt server is still work in progress.
<img width="663" alt="image" src="https://user-images.githubusercontent.com/2464813/228988038-5cfad045-e73d-4d3e-8d07-0a0e082cc2cb.png">


Bloop/SBT/SBT Bloop/SBT server is confusing to users.

---

With resolvers; since we are in the JVM ecosystem, it would be great to simply be able to say 'use M2_HOME for the settings.xml for connecting to the mirror', and job is done. Currently, to resolve these things anywhere that does not have access to the public internet, you need a tenacious expert to make things work. First impressions really count.

I appreciate none of these are as exciting or as engaging as an architectural redesign, but these issues of simply getting started hamper adoption and user experience of Scala significantly.

The other day I had some proxy settings that would not go away, even by setting SBT_OPTS, and JAVA_OPTS and other options, but worked specifying directly onto `java -Dhttps.proxy= sbt-launch.jar`. Also SBT bootstrap configurations and so forth... not the easier thing in the world to understand.

Where SBT is really awesome is in its super fast incremental compilation, watch triggers, strongly typed yet dynamic and flexible configurations (especially ability to embed Scala but have it rejected by the macros if it doesn't make sense). Once you have SBT running, it beats Maven and Gradle absolutely hands down. But that's _once you have it running_ -- and the launching experience has been one of the trickiest beasts for a long time.

https://github.com/sbt/sbt/issues/3410 has been opened for nearly 6 years, but once it is completed (please can we have a release? 😶‍🌫️ ), it will be such a huge improvement on user experience, but there is a lot more to go.


----

In short, let's fix the launching and set-up experience. SBT is the best build tool out there, once you get it running.

**Reply by eed3si9n** [reply]
This discussion is for discussing the RFC process of sbt project, not what goes into sbt. If you have ideas to improve sbt that you'd like to contribute, please create a separate discussion thread.
